### ISSUE TITLE
Is it possible to decode with the best model?

### ISSUE THREAD

**Issue by bnr887** [tensorflow/tensor2tensor#1798]
Hello, I'm new to T2T and still trying to understand some concepts. I'd like to decode my input using the model which has the best loss on validation set. Is that possible? Does export_save_model do this? I found [this issue](https://github.com/tensorflow/tensor2tensor/pull/974) and "export=FLAGS.export_saved_model" in the t2t_trainer.py file.

However, by using export_saved_model during my training, I couldn't observe any difference (it looks like still the latest models are saved into the train directory)?

Can someone please explain me this?

Many thanks!

**Comment by martinpopel** [comment]
You can keep more than the default 5 checkpoints with `t2t-trainer  --keep_checkpoint_max=N` (other relevant options are `--keep_checkpoint_every_n_hours` and `--save_checkpoints_secs`). You can then select the optimal checkpoint yourself (according to validation loss, or any other metric). Exporting is needed only for deploying the selected model (checkpoint) on server, but you can use `t2t-decode` with unexported models.

**Comment by bnr887** [comment]
Thank you so much for your explanations @martinpopel! Is there also a parameter for early stopping? I want to stop training as soon as the loss starts increasing on the validation set. Or should I manually check on tensorboard?

**Comment by martinpopel** [comment]
I'm not aware of automatic early-stopping options in T2T, but I may be wrong (I don't use the newest versions). t2t-trainer does not do proper autoregressive validation by default, but there are several other reasons why I don't use automatic early-stopping even in tools (such as [Marian](https://marian-nmt.github.io/docs/cmd/marian/)), which have these options (`--early-stopping --valid-freq --valid-metrics --keep-best` etc.).

**Comment by bnr887** [comment]
I see @martinpopel, thanks again!

**Comment by bnr887** [comment]
Hello, sorry to open this again! I'm working on summarization and I realized that although loss increases on validation after a certain number of training steps, accuracy and and accuracy_per_sequence also increase. How to interpret this?

Also, in this [example](https://github.com/GoogleCloudPlatform/training-data-analyst/blob/master/courses/machine_learning/deepdive/09_sequence/poetry.ipynb), the tuning objective is set to accuracy_per_sequence. Would it make sense to consider accuracy_per_sequence metric instead of loss to get the best model in my case?
